provisioning.go

```diff
@@ -0,0 +1,370 @@
+		}
+		return xerrors.Errorf("cannot create storage root key: %w", err)
+	}
+	defer tpm.FlushContext(transientSrkContext)
+
+	srkContext, err = tpm.EvictControl(tpm.OwnerHandleContext(), transientSrkContext, srkHandle, session)
+	if err != nil {
+		// Owner auth failure would have been caught by CreatePrimary
+		return xerrors.Errorf("cannot make storage root key persistent: %w", err)
+	}
+	tpm.provisionedSrk = srkContext
+
+	// Provision a lock NV index
+	if err := ensureLockNVIndex(tpm.TPMContext, session); err != nil {
+		if isNVIndexDefinedWithHandleError(err) {
+			return TPMResourceExistsError{err.(*nvIndexDefinedError).handle}
+		}
+		return xerrors.Errorf("cannot create lock NV index: %w", err)
+	}
+
+	if mode == ProvisionModeWithoutLockout {
+		return nil
+	}
+
+	// Perform actions that require the lockout hierarchy authorization.
+
+	// Set the DA parameters.
+	if err := tpm.DictionaryAttackParameters(tpm.LockoutHandleContext(), maxTries, recoveryTime, lockoutRecovery, session); err != nil {
+		switch {
+		case isAuthFailError(err):
+			return AuthFailError{tpm2.HandleLockout}
+		case isLockoutError(err):
+			return ErrTPMLockout
+		}
+		return xerrors.Errorf("cannot configure dictionary attack parameters: %w", err)
+	}
+
+	// Disable owner clear
+	if err := tpm.ClearControl(tpm.LockoutHandleContext(), true, session); err != nil {
+		// Lockout auth failure or lockout mode would have been caught by DictionaryAttackParameters
+		return xerrors.Errorf("cannot disable owner clear: %w", err)
+	}
+
+	// Set the lockout hierarchy authorization.
+	if err := tpm.HierarchyChangeAuth(tpm.LockoutHandleContext(), tpm2.Auth(newLockoutAuth),
+		session.IncludeAttrs(tpm2.AttrCommandEncrypt)); err != nil {
+		return xerrors.Errorf("cannot set the lockout hierarchy authorization value: %w", err)
+	}
+
+	return nil
+}
+
+// RequestTPMClearUsingPPI submits a request to the firmware to clear the TPM on the next reboot. This is the only way to clear
+// the TPM if owner clear has been disabled for the TPM, or the lockout hierarchy authorization value has been set previously but
+// is unknown.
+func RequestTPMClearUsingPPI() error {
+	f, err := os.OpenFile(ppiPath, os.O_WRONLY, 0)
+	if err != nil {
+		return xerrors.Errorf("cannot open request handle: %w", err)
+	}
+	defer f.Close()
+
+	if _, err := f.WriteString(clearPPIRequest); err != nil {
+		return xerrors.Errorf("cannot submit request: %w", err)
+	}
+
+	return nil
+}
+
+// ProvisionStatus returns the provisioning status for the specified TPM.
+func ProvisionStatus(tpm *TPMConnection) (ProvisionStatusAttributes, error) {
+	var out ProvisionStatusAttributes
+
+	session := tpm.HmacSession().IncludeAttrs(tpm2.AttrAudit)
+
+	if ek, err := tpm.CreateResourceContextFromTPM(ekHandle, session); err != nil {
+		if _, unavail := err.(tpm2.ResourceUnavailableError); !unavail {
+			return 0, err
+		}
+	} else if ekInit, err := tpm.EndorsementKey(); err == nil && bytes.Equal(ekInit.Name(), ek.Name()) {
```

Comment:
I'm not a fan of these nested if/else as I feel it requires to keep a lot of state in the head. I wonder if it could be written more explicit. Something like (please double check my logic!):
```
func isResourceUnavailable(err error) bool {
	_, unavail := err.(tpm2.ResourceUnavailableError)
	return unavail
}
...
        ek, err := tpm.CreateResourceContextFromTPM(ekHandle, session)
	if err != nil && !isResourceUnavailable(err) {
		return 0, err
	}
	if isResourceUnavailable(err) {
		ekInit, err := tpm.EndorsementKey()
		if err != nil {
			// yes?
			return err
		}
		if bytes.Equal(ekInit.Name(), ek.Name()) {
			out |= AttrValidEK
		}
        }
    ...
```

Reply:
The issue is that whilst we want to return early with an error if TPMContext.CreateResourceContextFromTPM returns an error that isn't tpm2.ResourceUnavailableError, we have a subsequent block of code that should only run if no error was returned. So, it would look more like this:
```
	ek, err := tpm.CreateResourceContextFromTPM(ekHandle, session)
	if err != nil && !isResourceUnavailable(err) {
		return 0, err
	}
	if err == nil {
		ekInit, err := tpm.EndorsementKey()
		if err == nil && bytes.Equal(ekInit.Name(), ek.Name()) {
			out |= AttrValidEK
		}
        }
```
Is that better though?

Reply:
Aha, sorry, again sorry if I'm too much bikesheeding, I will try to avoid this, maybe one more suggestion?
```
ek, err := tpm.CreateResourceContextFromTPM(ekHandle, session)
	switch {
	case err != nil && !isResourceUnavailable(err):
		// pass real error up
		return 0, err
	case isResourceUnavailable(err):
		// nothing to do
	case err == nil:
		ekInit, err := tpm.EndorsementKey()
		if err == nil && bytes.Equal(ekInit.Name(), ek.Name()) {
			out |= AttrValidEK
		}
	}
```
I like the very explicit nature of this. But feel free to ignore at this point it's probably more about personal preference (of me) than real improvements.